src/java.base/share/classes/jdk/crac/impl/CriticalUnorderedContext.java

```diff
@@ -0,0 +1,90 @@
+package jdk.crac.impl;
+
+import jdk.crac.CheckpointException;
+import jdk.crac.Context;
+import jdk.crac.Resource;
+import jdk.crac.RestoreException;
+
+import java.util.List;
+import java.util.Map;
+import java.util.WeakHashMap;
+
+/**
+ * Context performing Checkpoint notification in unspecified order.
+ * Concurrent registration along beforeCheckpoint notification triggers
+ * immediate notification on being registered resource.
+ * @param <R>
+ */
+public class CriticalUnorderedContext<R extends Resource> extends AbstractContext<R> {
+    private final WeakHashMap<R, Void> resources = new WeakHashMap<>();
+    private ExceptionHolder<CheckpointException> concurrentCheckpointException = null;
+
+    private synchronized List<R> snapshot() {
+        return this.resources.entrySet().stream()
+            .map(Map.Entry::getKey)
+            .toList();
+    }
+
+    @Override
+    protected List<R> checkpointSnapshot() {
+        return snapshot();
+    }
+
+    @Override
+    protected List<R> restoreSnapshot() {
+        // return updated set, as registration has called beforeCheckpoint()
+        return snapshot();
+    }
+
+    @Override
+    public void beforeCheckpoint(Context<? extends Resource> context) throws CheckpointException {
+        synchronized (this) {
+            concurrentCheckpointException = new ExceptionHolder<>(CheckpointException::new);
+        }
+
+        try {
+            super.beforeCheckpoint(context);
+        } catch (CheckpointException e) {
+            synchronized (this) {
+                concurrentCheckpointException.handle(e);
+            }
+        }
+        synchronized (this) {
+            ExceptionHolder<CheckpointException> e = concurrentCheckpointException;
+            concurrentCheckpointException = new ExceptionHolder<>(CheckpointException::new);
+            e.throwIfAny();
+        }
+    }
+
+    @Override
+    public void afterRestore(Context<? extends Resource> context) throws RestoreException {
+        CheckpointException racedException;
+        synchronized (this) {
+            racedException = concurrentCheckpointException.getIfAny();
+            concurrentCheckpointException = null;
+        }
+
+        ExceptionHolder<RestoreException> restoreException = new ExceptionHolder<>(RestoreException::new);
+        restoreException.handle(racedException);
+
+        restoreException.runWithHandler(() -> {
+            super.afterRestore(context);
+        });
+
+        restoreException.throwIfAny();
+    }
+
+    @Override
+    public void register(R resource) {
+        synchronized (this) {
+            resources.put(resource, null);
+            if (concurrentCheckpointException != null) {
+                try {
+                    invokeBeforeCheckpoint(resource);
```

Comment:
This context is blocking as well - since you hold the monitor on this object if the code tries to indirectly (via operation in another thread) register other resource, it will deadlock, and cannot be woken up via interrupt.
If the only usage is call sites we know that there won't be a recursive registration, so this not such a big flaw, though.

Reply:
Right, this context implementation something that should suit JDK needs, and the point is to verify the user is capable to do something similar with existing API. So, indeed, there is a possiblity of the deadlock with other thread. Fixing this does not seem trivial, so let's assume JDK will have to do something about this, if the problem become possible.